README.md
# raspistillWeb

Small Python web interface for raspistill to take photos with the RaspberryPi in the browser. I used the [pyramid framework](http://www.pylonsproject.org/) for site generation and [bootstrap](http://getbootstrap.com/) for fancy mobile-first layouts. 

![alt text](https://raw.github.com/TimJuni/raspistillWeb/master/raspistillweb/pictures/preview.jpg "raspistillWeb preview")

## Requirements

For a successful installation you need:
* a RaspberryPi with [raspbian](http://www.raspbian.org) installed (other os not tested, but should work too)
* a RaspberryPi Camera Kit

## Installation Notes

I'll provide a guide to install raspistillWeb based on the tutorial from the [pyramid framework](http://docs.pylonsproject.org/projects/pyramid/en/1.4-branch/narr/install.html) to install raspistillWeb in a seperate python environment, so that your systems python environment is not effected.

1. Make sure that your raspbian and your camera is working. Try to make a photo with raspistill to verify that your camera is working. 

2. Install python3-dev (if not already on your system), setuptools and exif and virtualenv via pip:
  * `sudo apt-get install python3-dev python3-pip python3-setuptools exif`
  * `sudo pip3 install virtualenv`

3. Create a virtual environment for python (sudo not required and not recommended):
  * `mkdir ~/Development` (Or another directory)
  * `cd ~/Development`
  * `virtualenv --python=python3 env`
  * `cd env`

4. Install raspistillWeb
  * `git clone https://github.com/TimJuni/raspistillWeb.git`
  * `cd raspistillWeb`
  * `../bin/python3 setup.py develop` (this may take some time)

5. Initialize the database
  * `../bin/initialize_raspistillweb_db development.ini`

6. Run raspistillWeb
  * `../bin/pserve development.ini`
  * surf `http://<adress of your pi>:6543`

## Updates
To update your Version of raspistillWeb, simply go into the raspistillWeb Directory and type:
  * `git pull`
  * `../bin/python setup.py develop`

Please notice, that you may need to install new dependencies (for example exif via apt-get)

## Example Configurations

### Running raspistillWeb over ssh
If you run raspistillWeb over ssh, you cannot quit your ssh session, because the raspistillWeb process will be killed. One easy way to start raspistillWeb (and let it run even when you are logged off) is to use `screen`:

1. Install screen
  * `sudo apt-get install screen`
2. Create a new screen
  * `screen -S cam`
3. Start raspistillWeb
  * `../bin/pserve development.ini`
4. Detatch the screen
  * simply hit ctrl + a + d 

At this point you can quit your ssh session and raspistillWeb will continue serving files. If you later want stop raspistillWeb, you can attach to the screen by: `screen -r cam`

### Running raspistillWeb behind a Reverse Proxy (with Authentification)
If you want to use to raspistillWeb from the internet, you can set up an apache reverse proxy to restrict the access to the application. Therefore you have to open port 80 in your router for your pi and follow these instructions:


1. Install Apache Web Server
    * `sudo apt-get install apache2`
2. Enable Apache Proxy Mod
    * `sudo a2enmod proxy_http`
3. Disable Default Site
    * `sudo a2dissite default`
4. Create a new Virtualhost
    * `sudo nano /etc/apache2/sites-available/raspistillWeb` and add the following content:
    
   
```
<VirtualHost *:80>
  ServerName mydomain.org
 
  <Location />
    AuthType Basic
    AuthName "raspistillWeb Login"
    AuthUserFile /usr/local/passwd/passwords
    Require user guest
  </Location>

  ProxyPreserveHost On
  ProxyRequests off
  ProxyPass / http://localhost:6543/
  ProxyPassReverse / http://localhost:6543/
 
</VirtualHost>
``` 

5. Create a password file
    * `sudo mkdir /usr/local/passwd`
    * `sudo htpasswd -c /usr/local/passwd/passwords guest`
    * Enter a password
    
6. Enable the Virtualhost and restart the Webserver
    * `sudo a2ensite raspistillWeb`
    * `sudo /etc/init.d/apache2 restart`
    
When you surf `http://<adress of your pi>` you should be asked for a username (guest) and a password (see step 5). 

### Run raspistillWeb on startup
1. Create a startup script
  * `sudo nano /etc/init.d/raspistillweb`
```
#!/bin/bash
cd /home/pi/Development/env/raspistillWeb
../bin/pserve development.ini
``` 
2. Make the script executable
  `sudo chmod 755 /etc/init.d/raspistillweb`
3. Register the script to be run at startup
  `sudo update-rc.d /etc/init.d/raspistillweb defaults`

## Future Work
This is my first project with python and pyramid. Feel free to leave a commend or a feature request.


CHANGES.txt
0.2
---

- Added timelapse support
- Added persistent storage of settings and pictures


0.1
---

- Added Archive


0.0
---

-  Initial version


raspistillweb/tests.py
# raspistillWeb - web interface for raspistill
# Copyright (C) 2013 Tim Jungnickel
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.

import unittest

from pyramid import testing


class ViewTests(unittest.TestCase):
    def setUp(self):
        self.config = testing.setUp()

    def tearDown(self):
        testing.tearDown()



raspistillweb/models.py
from sqlalchemy import (
    Column,
    Index,
    Integer,
    Text,
    )

from sqlalchemy.ext.declarative import declarative_base

from sqlalchemy.orm import (
    scoped_session,
    sessionmaker,
    )

from zope.sqlalchemy import ZopeTransactionExtension

DBSession = scoped_session(sessionmaker(extension=ZopeTransactionExtension()))
Base = declarative_base()


class Picture(Base):
    __tablename__ = 'models'
    id = Column(Integer, primary_key=True)
    filename = Column(Text)
    image_effect = Column(Text)
    exposure_mode = Column(Text)
    awb_mode = Column(Text)
    resolution = Column(Text)
    ISO = Column(Integer)
    exposure_time = Column(Text)
    date = Column(Text)
    timestamp = Column(Text)
    filesize = Column(Integer)

class Settings(Base):
    __tablename__ = 'settings'
    id = Column(Integer, primary_key=True)
    image_width = Column(Integer)
    image_height = Column(Integer)
    timelapse_interval = Column(Integer)
    timelapse_time = Column(Integer)
    exposure_mode = Column(Text)
    image_effect = Column(Text)
    awb_mode = Column(Text)
    image_ISO = Column(Text)
    image_rotation = Column(Text)

class Timelapse(Base):
    __tablename__ = 'timelapse'
    id = Column(Integer, primary_key=True)
    filename = Column(Text)
    timeStart = Column(Text)
    image_effect = Column(Text)
    exposure_mode = Column(Text)
    awb_mode = Column(Text)
    timeEnd = Column(Text)

Index('my_index', Picture.filename, unique=True, mysql_length=255)


raspistillweb/__init__.py
# raspistillWeb - web interface for raspistill
# Copyright (C) 2013 Tim Jungnickel
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.

from pyramid.config import Configurator
from sqlalchemy import engine_from_config

from .models import (
    DBSession,
    Base,
    )

def main(global_config, **settings):
    """ This function returns a Pyramid WSGI application.
    """
    engine = engine_from_config(settings, 'sqlalchemy.')
    DBSession.configure(bind=engine)
    Base.metadata.bind = engine
    
    config = Configurator(settings=settings)
    #config.include('pyramid_chameleon')
    config.include('pyramid_mako')
    config.add_static_view('static', 'static', cache_max_age=3600)
    config.add_static_view('pictures', 'pictures', cache_max_age=3600)
    config.add_static_view('thumbnails', 'thumbnails', cache_max_age=3600)
    config.add_static_view('time-lapse', 'time-lapse', cache_max_age=3600)
    config.add_route('home','/')
    config.add_route('settings','/settings')
    config.add_route('save','/save')
    config.add_route('delete_picture','/delete_picture')
    config.add_route('delete_timelapse','/delete_timelapse')
    config.add_route('archive','/archive')
    config.add_route('timelapse','/timelapse')
    config.add_route('photo','/photo')
    config.add_route('timelapse_start','/timelapse_start')
    config.scan()
    return config.make_wsgi_app()


raspistillweb/scripts/initializedb.py
import os
import sys
import transaction

from sqlalchemy import engine_from_config

from pyramid.paster import (
    get_appsettings,
    setup_logging,
    )

from pyramid.scripts.common import parse_vars

from ..models import (
    DBSession,
    Picture,
    Settings,
    Timelapse,
    Base,
    )


def usage(argv):
    cmd = os.path.basename(argv[0])
    print('usage: %s <config_uri> [var=value]\n'
          '(example: "%s development.ini")' % (cmd, cmd))
    sys.exit(1)


def main(argv=sys.argv):
    if len(argv) < 2:
        usage(argv)
    config_uri = argv[1]
    options = parse_vars(argv[2:])
    setup_logging(config_uri)
    settings = get_appsettings(config_uri, options=options)
    engine = engine_from_config(settings, 'sqlalchemy.')
    DBSession.configure(bind=engine)
    Base.metadata.create_all(engine)
    with transaction.manager:
        picture = Picture(
                        filename = '1',
                        image_effect = '1',
                        exposure_mode = '1',
                        awb_mode = '1',
                        resolution = '1',
                        ISO = 1,
                        exposure_time ='1',
                        date = '1',
                        timestamp = '1',
                        filesize = '1'
                        )
        DBSession.add(picture)

        app_settings = Settings(
                        image_width = 800,
                        image_height = 600,
                        timelapse_interval = 4000,
                        timelapse_time = 20000,
                        exposure_mode = 'auto',
                        image_effect = 'none',
                        awb_mode = 'auto',
                        image_ISO = 'auto',
                        image_rotation = '0'
                        )
        DBSession.add(app_settings)

        timelapse = Timelapse(
                        filename = 'test',
                        timeStart = 'test',
                        image_effect = 'test',
                        exposure_mode = 'test',
                        awb_mode = 'auto',
                        timeEnd = 'none'
                        )

        DBSession.add(timelapse)


raspistillweb/views.py
# raspistillWeb - web interface for raspistill
# Copyright (C) 2013 Tim Jungnickel
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.

from pyramid.view import view_config
from pyramid.httpexceptions import HTTPFound
import exifread
import os
import threading
import tarfile
from subprocess import call
from time import gmtime, strftime, localtime, asctime, mktime
from stat import *
from datetime import *

from sqlalchemy.exc import DBAPIError

import transaction

from .models import (
    DBSession,
    Picture,
    Settings,
    Timelapse,
    )


# Modify these lines to change the directory where the pictures and thumbnails
# are stored. Make sure that the directories exist and the user who runs this
# program has write access to the directories. 
RASPISTILL_DIRECTORY = 'raspistillweb/pictures/' # Example: /home/pi/pics/
THUMBNAIL_DIRECTORY = 'raspistillweb/thumbnails/' # Example: /home/pi/thumbs/
TIMELAPSE_DIRECTORY = 'raspistillweb/time-lapse/' # Example: /home/pi/timelapse/

IMAGE_EFFECTS = [
    'none', 'negative', 'solarise', 'sketch', 'denoise', 'emboss', 'oilpaint', 
    'hatch', 'gpen', 'pastel', 'watercolour', 'film', 'blur', 'saturation', 
    'colourswap', 'washedout', 'posterise', 'colourpoint', 'colourbalance', 
    'cartoon'
    ]

EXPOSURE_MODES = [
    'auto', 'night', 'nightpreview', 'backlight', 'spotlight', 'sports',
    'snow', 'beach', 'verylong', 'fixedfps', 'antishake', 'fireworks'
    ]
    
AWB_MODES = [
    'off', 'auto', 'sun', 'cloud', 'shade', 'tungsten', 'fluorescent',
    'incandescent', 'flash', 'horizon'
    ]

ISO_OPTIONS = [
    'auto', '100', '150', '200', '250', '300', '400', '500', 
    '600', '700', '800'
    ]
    
IMAGE_RESOLUTIONS = [
    '800x600','1024x786','1900x1200','1280x720','1920x1080', '2593x1944'
    ]
    
IMAGE_HEIGHT_ALERT = 'Please enter an image height between 0 and 1945.'
IMAGE_WIDTH_ALERT = 'Please enter an image width between 0 and 2593.'
IMAGE_EFFECT_ALERT = 'Please enter a valid image effect.'
EXPOSURE_MODE_ALERT = 'Please enter a valid exposure mode.'
AWB_MODE_ALERT = 'Please enter a valid awb mode.'
ISO_OPTION_ALERT = 'Please enter a valid ISO option.'
IMAGE_ROTATION_ALERT = 'Please enter a valid image rotation option.'

THUMBNAIL_SIZE = '240:160:80'

timelapse = False
timelapse_database = None

preferences_fail_alert = []
preferences_success_alert = False

# not implemented yet
image_quality = '100'
image_sharpness = '0'
image_contrast = '0'
image_brightness = '50'
image_saturation = '0'


###############################################################################
################################### Views #####################################
###############################################################################


# View for the /settings site
@view_config(route_name='settings', renderer='settings.mako')
def settings_view(request):
    app_settings = DBSession.query(Settings).first()
    global preferences_fail_alert, preferences_success_alert
        
    preferences_fail_alert_temp = []    
    if preferences_fail_alert is not []:
        preferences_fail_alert_temp = preferences_fail_alert
        preferences_fail_alert = []
     
    preferences_success_alert_temp = False  
    if preferences_success_alert:
        preferences_success_alert_temp = True
        preferences_success_alert = False
        
    return {'project' : 'raspistillWeb',
            'image_effect' : app_settings.image_effect,
            'exposure_mode' : app_settings.exposure_mode,
            'awb_mode' : app_settings.awb_mode,
            'image_effects' : IMAGE_EFFECTS,
            'exposure_modes' : EXPOSURE_MODES,
            'awb_modes' : AWB_MODES,
            'image_width' : str(app_settings.image_width),
            'image_height' : str(app_settings.image_height),
            'image_iso' : app_settings.image_ISO,
            'iso_options' :  ISO_OPTIONS, 
            'timelapse_interval' : str(app_settings.timelapse_interval),
            'timelapse_time' : str(app_settings.timelapse_time),
            'preferences_fail_alert' : preferences_fail_alert_temp,
            'preferences_success_alert' : preferences_success_alert_temp,
            'image_rotation' : app_settings.image_rotation,
            'image_resolutions' : IMAGE_RESOLUTIONS
            } 
            
# View for the /archive site
@view_config(route_name='archive', renderer='archive.mako')
def archive_view(request):

    pictures = DBSession.query(Picture).all()
    picturedb = []
    for picture in pictures:
        imagedata = get_picture_data(picture)
        picturedb.insert(0,imagedata)
    return {'project' : 'raspistillWeb',
            'database' : picturedb
            }
                    
# View for the / site
@view_config(route_name='home', renderer='home.mako')
def home_view(request):
    pictures = DBSession.query(Picture).all()
    if len(pictures) == 0:
        return HTTPFound(location='/photo')
    else:
        picture_data = get_picture_data(pictures[-1])
        if timelapse:            
            return {'project': 'raspistillWeb',
                    'imagedata' : picture_data,
                    'timelapse' : timelapse,
                    }
        #elif (mktime(localtime()) - mktime(picture_data['timestamp'])) > 1800: 
        #    return HTTPFound(location='/photo') 
        else:
            return {'project': 'raspistillWeb',
                    'imagedata' : picture_data,
                    'timelapse' : timelapse,
                    }

# View for the /timelapse site        
@view_config(route_name='timelapse', renderer='timelapse.mako')
def timelapse_view(request):
    global timelapse_database

    #if timelapse_database is not None:
    #    DBSession.add(timelapse_database)
    #    timelapse_database = None

    app_settings = DBSession.query(Settings).first()
    timelapse_collection = DBSession.query(Timelapse).all()
    timelapsedb = []
    for timelapse_rec in timelapse_collection:
        timelapse_data = get_timelapse_data(timelapse_rec)
        timelapsedb.insert(0,timelapse_data)
    
    return {'project': 'raspistillWeb',
            'timelapse' : timelapse,
            'timelapseInterval' : str(app_settings.timelapse_interval),
            'timelapseTime' : str(app_settings.timelapse_time),
            'timelapseDatabase' : timelapsedb
            }
            
# View for the timelapse start - no site will be generated
@view_config(route_name='timelapse_start')
def timelapse_start_view(request):
    global timelapse
    timelapse = True
    filename = strftime("%Y-%m-%d.%H.%M.%S", localtime())
    t = threading.Thread(target=take_timelapse, args=(filename, ))
    t.start()
    return HTTPFound(location='/timelapse') 


# View to take a photo - no site will be generated
@view_config(route_name='photo')
def photo_view(request):
    if timelapse:
        return HTTPFound(location='/') 
    else:
        app_settings = DBSession.query(Settings).first()
        filename = strftime("%Y-%m-%d.%H.%M.%S.jpg", localtime())
        take_photo(filename)
        
        f = open(RASPISTILL_DIRECTORY + filename,'rb')
        exif = extract_exif(exifread.process_file(f))    
        filedata = extract_filedata(os.stat(RASPISTILL_DIRECTORY + filename))
        filedata.update(exif)
        filedata['filename'] = filename
        filedata['image_effect'] = app_settings.image_effect
        filedata['exposure_mode'] = app_settings.exposure_mode
        filedata['awb_mode'] = app_settings.awb_mode
        '''
        imagedata = dict()
        imagedata['filename'] = filename
        imagedata['image_effect'] = 'test'
        imagedata['exposure_mode'] = 'test'
        imagedata['awb_mode'] = 'test'
        imagedata['resolution'] = '800x600'
        imagedata['ISO'] = '300'
        imagedata['exposure_time'] = '100'
        imagedata['date'] = 'test'
        imagedata['timestamp'] = localtime()
        imagedata['filesize'] = 100
        '''
        picture = Picture(filename=filedata['filename'],
                        image_effect=filedata['image_effect'],
                        exposure_mode=filedata['exposure_mode'],
                        awb_mode=filedata['awb_mode'],
                        resolution=filedata['resolution'],
                        ISO=filedata['ISO'],
                        exposure_time=filedata['exposure_time'],
                        date=filedata['date'],
                        timestamp='test',
                        filesize=filedata['filesize'])
        DBSession.add(picture)
        return HTTPFound(location='/')  
         
# View for the archive delete - no site will be generated
@view_config(route_name='delete_picture')
def pic_delete_view(request):    
    p_id = request.params['id']
    pic = DBSession.query(Picture).filter_by(id=int(p_id)).first()        
    DBSession.delete(pic)
    return HTTPFound(location='/archive')

# View for the timelapse delete - no site will be generated
@view_config(route_name='delete_timelapse')
def tl_delete_view(request):
    t_id = request.params['id']
    tl = DBSession.query(Timelapse).filter_by(id=int(t_id)).first()        
    DBSession.delete(tl)
    return HTTPFound(location='/timelapse')

# View for settings form data - no site will be generated      
@view_config(route_name='save')
def save_view(request):

    global preferences_success_alert, preferences_fail_alert

    image_width_temp = request.params['imageWidth']
    image_height_temp = request.params['imageHeight']
    timelapse_interval_temp = request.params['timelapseInterval']
    timelapse_time_temp = request.params['timelapseTime']
    exposure_mode_temp = request.params['exposureMode']
    image_effect_temp = request.params['imageEffect']
    awb_mode_temp = request.params['awbMode']
    image_ISO_temp = request.params['isoOption']
    image_rotation_temp = request.params['imageRotation']
    image_resolution = request.params['imageResolution']

    app_settings = DBSession.query(Settings).first()
    
    if image_width_temp:
        if 0 < int(image_width_temp) < 2593:
            app_settings.image_width = image_width_temp
        else:
            preferences_fail_alert.append(IMAGE_WIDTH_ALERT)
    
    if image_height_temp:
        if 0 < int(image_height_temp) < 1945:
            app_settings.image_height = image_height_temp
        else:
            preferences_fail_alert.append(IMAGE_HEIGHT_ALERT)
            
    if not image_width_temp and not image_height_temp:
        app_settings.image_width = image_resolution.split('x')[0]
        app_settings.image_height = image_resolution.split('x')[1]
            
    if timelapse_interval_temp:
        app_settings.timelapse_interval = timelapse_interval_temp
        
    if timelapse_time_temp:
        app_settings.timelapse_time = timelapse_time_temp
    
    if exposure_mode_temp and exposure_mode_temp in EXPOSURE_MODES:
        app_settings.exposure_mode = exposure_mode_temp
    else:
        preferences_fail_alert.append(EXPOSURE_MODE_ALERT)
        
    if image_effect_temp and image_effect_temp in IMAGE_EFFECTS:
        app_settings.image_effect = image_effect_temp
    else:
        preferences_fail_alert.append(IMAGE_EFFECT_ALERT)
        
    if awb_mode_temp and awb_mode_temp in AWB_MODES:
        app_settings.awb_mode = awb_mode_temp
    else:
        preferences_fail_alert.append(AWB_MODE_ALERT)
        
    if image_ISO_temp and image_ISO_temp in ISO_OPTIONS:
        app_settings.image_ISO = image_ISO_temp
    else:
        preferences_fail_alert.append(ISO_OPTION_ALERT)
        
    if image_rotation_temp and image_rotation_temp in ['0','90','180','270']:
        app_settings.image_rotation = image_rotation_temp
    else:
        preferences_fail_alert.append(IMAGE_ROTATION_ALERT)  
        
    if preferences_fail_alert == []:
        preferences_success_alert = True 
    
    DBSession.flush()      
    return HTTPFound(location='/settings')  

###############################################################################
############ Helper functions to keep the code clean ##########################
###############################################################################

def take_photo(filename):
    app_settings = DBSession.query(Settings).first()
    
    if app_settings.image_ISO == 'auto':
        iso_call = ''
    else:
        iso_call = ' -ISO ' + str(app_settings.image_ISO)
    call (
        ['raspistill -t 500'
        + ' -w ' + str(app_settings.image_width)
        + ' -h ' + str(app_settings.image_height)
        + ' -ex ' + app_settings.exposure_mode
        + ' -awb ' + app_settings.awb_mode
        + ' -rot ' + str(app_settings.image_rotation)
        + ' -ifx ' + app_settings.image_effect
        + iso_call
        + ' -th ' + THUMBNAIL_SIZE 
        + ' -o ' + RASPISTILL_DIRECTORY + filename], shell=True
        )
    if not (RASPISTILL_DIRECTORY == 'raspistillweb/pictures/'):
        call (
            ['ln -s ' + RASPISTILL_DIRECTORY + filename
            + ' raspistillweb/pictures/' + filename], shell=True
            )
    generate_thumbnail(filename)
    

    #call(['cp raspistillweb/pictures/preview.jpg raspistillweb/pictures/'+filename],shell=True)
    #generate_thumbnail(filename)
    return

    
def take_timelapse(filename):
    global timelapse, timelapse_database

    app_settings = DBSession.query(Settings).first()
    timelapsedata = {'filename' :  filename}
    timelapsedata['timeStart'] = str(asctime(localtime()))
    os.makedirs(TIMELAPSE_DIRECTORY + filename)
    call (
        ['raspistill'
        + ' -w ' + str(app_settings.image_width)
        + ' -h ' + str(app_settings.image_height)
        + ' -ex ' + app_settings.exposure_mode
        + ' -awb ' + app_settings.awb_mode
        + ' -ifx ' + app_settings.image_effect
        + ' -th ' + THUMBNAIL_SIZE
        + ' -tl ' + str(app_settings.timelapse_interval)
        + ' -t ' + str(app_settings.timelapse_time) 
        + ' -o ' + TIMELAPSE_DIRECTORY + filename + '/'
        + filename + '_%04d.jpg'], shell=True
        )    
    timelapsedata['image_effect'] = app_settings.image_effect
    timelapsedata['exposure_mode'] = app_settings.exposure_mode
    timelapsedata['awb_mode'] = app_settings.awb_mode
    timelapsedata['timeEnd'] = str(asctime(localtime()))

    timelapse_data = Timelapse(
                        filename = timelapsedata['filename'],
                        timeStart = timelapsedata['timeStart'],
                        image_effect = timelapsedata['image_effect'],
                        exposure_mode = timelapsedata['exposure_mode'],
                        awb_mode = timelapsedata['awb_mode'],
                        timeEnd = timelapsedata['timeEnd'],
                    )

    print('Adding timelapse to DB')
    DBSession.add(timelapse_data)
    #DBSession.flush() 
    transaction.commit()
    print('Added timelapse to DB')
    with tarfile.open(TIMELAPSE_DIRECTORY + filename + '.tar.gz', "w:gz") as tar:
        tar.add(TIMELAPSE_DIRECTORY + filename, arcname=os.path.basename(TIMELAPSE_DIRECTORY + filename))

    #timelapse_database = timelapse_data
    timelapse = False
    
    return 

def generate_thumbnail(filename):
    call (
        ['exif -e ' + RASPISTILL_DIRECTORY + filename
        + ' -o ' + THUMBNAIL_DIRECTORY + filename], shell=True
    )
    if not (THUMBNAIL_DIRECTORY == 'raspistillweb/thumbnails/'):
        call (
            ['ln -s ' + THUMBNAIL_DIRECTORY + filename 
            + ' raspistillweb/thumbnails/' + filename], shell=True
            )
    return

def extract_exif(tags):
    return {
        'resolution' : str(tags['Image ImageWidth']) 
        + ' x ' + str(tags['Image ImageLength']),
        'ISO' : str(tags['EXIF ISOSpeedRatings']),
        'exposure_time' : str(tags['EXIF ExposureTime'])
            }
         
def extract_filedata(st):
    return {
        'date' : str(asctime(localtime(st[ST_MTIME]))),
        'timestamp' : localtime(),
        'filesize': str((st[ST_SIZE])/1000) + ' kB'
            }
            
def get_picture_data(picture):
    imagedata = dict()
    imagedata['id'] = str(picture.id)
    imagedata['filename'] = picture.filename
    imagedata['image_effect'] = picture.image_effect
    imagedata['exposure_mode'] = picture.exposure_mode
    imagedata['awb_mode'] = picture.awb_mode
    imagedata['resolution'] = picture.resolution
    imagedata['ISO'] = str(picture.ISO)
    imagedata['exposure_time'] = picture.exposure_time
    imagedata['date'] = str(picture.date)
    imagedata['timestamp'] = str(picture.timestamp)
    imagedata['filesize'] = str(picture.filesize)
    return imagedata

def get_timelapse_data(timelapse_rec):
    timelapse_data = dict()
    timelapse_data['id'] = str(timelapse_rec.id)
    timelapse_data['filename'] = timelapse_rec.filename
    timelapse_data['image_effect'] = timelapse_rec.image_effect
    timelapse_data['exposure_mode'] = timelapse_rec.exposure_mode
    timelapse_data['awb_mode'] = timelapse_rec.awb_mode
    timelapse_data['timeStart'] = str(timelapse_rec.timeStart)
    timelapse_data['timeEnd'] = str(timelapse_rec.timeEnd)
    return timelapse_data
